server/game/core/card/UpgradeCard.ts

```diff
@@ -54,6 +54,47 @@ export class UpgradeCard extends UpgradeCardParent {
         this._parentCard = card as UnitCard;
     }
 
+    public override moveTo(targetLocation: Location) {
+        if (
+            !Contract.assertFalse(this._parentCard && targetLocation !== this._parentCard.location,
+                `Attempting to move upgrade ${this.internalName} while it is still attached to ${this._parentCard?.internalName}`)
+        ) {
+            return;
+        }
+
+        super.moveTo(targetLocation);
+    }
+
+    public attachTo(newParentCard: UnitCard) {
+        if (
+            !Contract.assertTrue(newParentCard.isUnit()) ||
+            !Contract.assertTrue(EnumHelpers.isArena(newParentCard.location))
+        ) {
+            return;
+        }
+
+        if (this._parentCard) {
+            this.unattach();
+        } else {
+            this.controller.removeCardFromPile(this);
+        }
+
+        this.moveTo(newParentCard.location);
+        newParentCard.controller.putUpgradeInArena(this, newParentCard.location);
+        newParentCard.attachUpgrade(this);
+        this._parentCard = newParentCard;
+    }
+
+    public unattach() {
```

Comment:
Does this method also handle moving to the discard, or does that happen elsewhere?

Reply:
That happens elsewhere, because it's not always being moved to discard after this.